### PR TITLE
refactor: combine Program.hints and Program.hints_ranges into custom collection

### LINKED ISSUE #1342: Refactor: combine `Program.hints` and `Program.hints_ranges` into a custom collection
Currently, the `Program` struct contains two fields `hints` and `hints_ranges`, where the second one is used to index the first.

https://github.com/lambdaclass/cairo-vm/blob/7e5378a8fd0c44718fcab523273334b7f75a27dc/vm/src/types/program.rs#L51-L53

As they are strongly related two each other, they should be combined into a single collection that encapsulates the creation and indexing of hints. Some related functionality would be:
- `Program::flatten_hints` (a kind of constructor for it),

https://github.com/lambdaclass/cairo-vm/blob/7e5378a8fd0c44718fcab523273334b7f75a27dc/vm/src/types/program.rs#L121-L124

- and indexing operations (they are used in more than one place)

https://github.com/lambdaclass/cairo-vm/blob/7e5378a8fd0c44718fcab523273334b7f75a27dc/vm/src/vm/runners/cairo_runner.rs#L589-L595

### PR DESCRIPTION
**Pull request by PanGan21**
# TITLE
refactor: combine Program.hints and Program.hints_ranges into custom collection

## Description
Combines `Program.hints` and `Program.hints_ranges` into a new struct called `hints_collection`.
Closes #1342

Description of the pull request changes and motivation.

## Checklist
- [x] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

File: vm/src/types/program.rs [hunk 1]
```diff
@@ -98,6 +95,12 @@ impl<'a> Arbitrary<'a> for SharedProgramData {
     }
 }
 
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct HintsCollection {
+    pub hints: Vec<HintParams>,
+    pub hints_ranges: Vec<HintRange>,
+}
```
Comment (Oppen) [inline vm/src/types/program.rs:102]: I'd suggest the following:
- The fields to be private;
- The struct to be pub(crate);
- Use a constructor taking a HashMap;
- Use a getter taking a range and returning a slice.

Comment (PanGan21) [reply vm/src/types/program.rs:102]: Hi! Thanks for the suggestions! Just a short question.
The getter you are describing works fine for this case:
```
let hint_data = self 
     .program 
     .shared_program_data 
     .hints_ranges 
     .get(vm.run_context.pc.offset) 
     .and_then(|r| r.and_then(|(s, l)| hint_data.get(s..s + l.get()))) 
     .unwrap_or(&[]); 
```

But cases like this:
```
self.program
            .shared_program_data
            .hints_collection
            .hints
            .iter()
            .map(|hint| {
                hint_executor
                    .compile_hint(
                        &hint.code,
                        &hint.flow_tracking_data.ap_tracking,
                        &hint.flow_tracking_data.reference_ids,
                        references,
                    )
                    .map_err(|_| VirtualMachineError::CompileHintFail(hint.code.clone().into()))
            })
            .collect()
```

It seems a getter that returns all the hints is required. Let me know if I am wrong :)

Comment (Oppen) [reply vm/src/types/program.rs:102]: You can make a method that returns the iterator instead:
```rust
fn iter(&self) -> impl Iterator<Item = (pc, &[HintData])> {
    self.iter()
}
```

Comment (Oppen) [reply vm/src/types/program.rs:102]: You could also implement some `fn iter_hints(&self) -> impl Iterator<Item = &HintData>` if it makes this easier.

Comment (PanGan21) [reply vm/src/types/program.rs]: Unfortunatelly I was made to implement both `iter_hints` and `iter`. The `iter` function is used only for testing in the helper `get_hints_as_map` thus it has `#[allow(dead_code)]`. Let me know your thoughts

File: vm/src/types/program.rs [hunk 2]
```diff
@@ -159,33 +227,9 @@ impl Program {
...
-                NonZeroUsize::new(hs.len()).expect("empty vecs already filtered"),
-            );
-            hints_ranges[*pc] = Some(range);
-            hints_values.extend_from_slice(&hs[..]);
-        }
-
-        Ok((hints_values, hints_ranges))
+    ) -> Result<HintsCollection, ProgramError> {
+        let hints_map: HashMap<usize, Vec<HintParams>> = hints.clone().into_iter().collect();
+        HintsCollection::new(hints_map, program_length)
```
[the remark this answers is not in the document]
Comment (Oppen) [reply vm/src/types/program.rs:232]: Oh, that was my bad, I forgot about it and recommended the HashMap.

File: vm/src/types/program.rs (hunk 1 above)
Comment (Oppen) [reply vm/src/types/program.rs]: You can avoid the `#[allow(dead_code)]` by using a separate `impl` block that is gated by the `#[cfg(test)]` annotation.

File: vm/src/types/program.rs [hunk 3]
```diff
@@ -98,6 +95,78 @@ impl<'a> Arbitrary<'a> for SharedProgramData {
...
+        })
+    }
+
+    pub fn iter_hints(&self) -> impl Iterator<Item = &HintParams> {
+        self.hints.iter()
+    }
+
+    pub fn get_hint_range_for_pc(&self, pc: usize) -> Option<HintRange> {
+        self.hints_ranges.get(pc).cloned()
+    }
```
Comment (Oppen) [inline vm/src/types/program.rs:150]: What I was thinking of was something more like:
```suggestion
    pub fn get_hint_range_for_pc(&self, pc: usize) -> Option<&[HintParams]> {
        let bounds = self.hints_ranges.get(pc)?;
        self.hints.get(bounds.0..bounds.0+bounds.1)
    }
```
This is directly usable later on.

File: vm/src/vm/runners/cairo_runner.rs [hunk 4]
```diff
@@ -528,6 +528,144 @@ impl CairoRunner {
...
+        //                     references,
+        //                 )?;
+
+        //                 hint_data.push(Box::new(compiled_hint) as Box<dyn Any>);
+
+        //                 Ok(())
+        //             })
+        //     })?;
+
+        // Ok(hint_data)
```
Comment (Oppen) [inline vm/src/vm/runners/cairo_runner.rs:668]: I gather this came from iterating solutions and slipped. Could you remove it before merge?

File: vm/src/types/program.rs (hunk 1 above)
Comment (PanGan21) [reply vm/src/types/program.rs]: Added the new impl block! Thanks!

File: vm/src/types/program.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (PanGan21) [reply vm/src/types/program.rs]: Fixed it and replaced `flatten_hints` with the constructor

File: vm/src/vm/runners/cairo_runner.rs (hunk 4 above)
Comment (PanGan21) [reply vm/src/vm/runners/cairo_runner.rs]: Oops. Good catch. Removed!

**Review by Oppen — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**